astoria-tech/volunteer-dispatch
# Send followup SMS to volunteers upon task completion

This PR adds the Twilio SMS API library, in order to send volunteers a "Thank You" text message upon task completion.

### Airtable Schema Updates
Two new fields are created in the `Requests` table:

- `Followup SMS Sent?`: tracks whether the volunteer has been sent a followup text upon task completion.
> ![Screen Shot 2020-07-01 at 7 18 54 PM](https://user-images.githubusercontent.com/6953434/86300713-c1ab2800-bbd1-11ea-9c0e-9d86fa72ad32.png)

- `Last modified time`: tracks when the `Status` field was last updated. This is necessary to select only `request` records that have been marked `Completed` in the last day, so as to avoid sending a text to all volunteers who have completed a task ever and possibly overloading the system.
> ![Screen Shot 2020-07-01 at 7 20 21 PM](https://user-images.githubusercontent.com/6953434/86300728-cf60ad80-bbd1-11ea-809a-59b66df94ab8.png)

### Dependencies Updates
Two new dependencies are added:
- `twilio`: Twilio API library
- `phone`: library to normalize mobile numbers into [E.164 format](https://www.twilio.com/docs/glossary/what-e164), as required by Twilio.

### Node Environment Variable Updates
Three new environment variables are added:

- `TWILIO_ACCOUNT_SID`
- `TWILIO_AUTH_TOKEN`
- `TWILIO_PHONE_NUMBER`

### Testing 
In order to test the text message functionality, follow these steps:
- [Verify your personal mobile number (or a number for testing) with Twilio](https://www.twilio.com/docs/usage/tutorials/how-to-use-your-free-trial-account#verify-your-personal-phone-number)
- In your copy of Airtable, set the value of `Followup SMS Sent?` to `Yes` for all records with a `Status` of `Completed`
- Choose the most recently completed `request` record, and update the corresponding volunteer's phone number in the `volunteers` table with your personal/testing number
- If the most recently completed `request` was completed more than a day ago, change the value of `Status` to one of the other options, then change it back to `Completed`. This will update `Last modified time`
- Set `Followup SMS Sent?` to `No` for the most recently completed `request` record

#### Console
> ![Screen Shot 2020-07-01 at 7 48 16 PM](https://user-images.githubusercontent.com/6953434/86301657-62024c00-bbd4-11ea-975f-5d95f71942c0.png)

#### Followup SMS
> ![Screenshot_2020-07-01-19-49-15](https://user-images.githubusercontent.com/6953434/86301739-a0980680-bbd4-11ea-992a-00d00b7d9887.png)
